desc/integrals/_interp_utils.py

```diff
@@ -0,0 +1,754 @@
+#   Let spectral resolution be F, (e.g. F = M N for 2D transform),
+#   and number of points (non-uniform) to evaluate be Q. A non-uniform
+#   fast transform cost is 𝒪([F+Q] log[F] log[1/ε]) where ε is the
+#   interpolation error term (depending on implementation how ε appears
+#   may change, but it is always logarithmic). Direct evaluation is 𝒪(F Q).
+#   Note that for the inverse Chebyshev transforms, we can also use fast
+#   multipoint methods Chapter 10, https://doi.org/10.1017/CBO9781139856065.
+#   Unlike NFFTs, multipoint methods are exact and reduce to using FFTs.
+#   The cost is 𝒪([F+Q] log²[F + Q]). This might be useful to evaluating
+#   |B|, since the integrands are not smooth functions of |B|, which we know
+#   as a Chebyshev series, and the nodes are packed more tightly near the
+#   singular regions.
+
+
+def interp_rfft(xq, f, domain=(0, 2 * jnp.pi), axis=-1):
+    """Interpolate real-valued ``f`` to ``xq`` with FFT.
+
+    Parameters
+    ----------
+    xq : jnp.ndarray
+        Real query points where interpolation is desired.
+        Shape of ``xq`` must broadcast with arrays of shape ``np.delete(f.shape,axis)``.
+    f : jnp.ndarray
+        Real function values on uniform grid over an open period to interpolate.
+    domain : tuple[float]
+        Domain over which samples were taken.
+    axis : int
+        Axis along which to transform.
+
+    Returns
+    -------
+    fq : jnp.ndarray
+        Real function value at query points.
+
+    """
+    a = rfft(f, axis=axis, norm="forward")
+    return irfft_non_uniform(xq, a, f.shape[axis], domain, axis)
+
+
+def irfft_non_uniform(xq, a, n, domain=(0, 2 * jnp.pi), axis=-1):
+    """Evaluate Fourier coefficients ``a`` at ``xq``.
+
+    Parameters
+    ----------
+    xq : jnp.ndarray
+        Real query points where interpolation is desired.
+        Shape of ``xq`` must broadcast with arrays of shape ``np.delete(a.shape,axis)``.
+    a : jnp.ndarray
+        Fourier coefficients ``a=rfft(f,axis=axis,norm="forward")``.
+    n : int
+        Spectral resolution of ``a``.
+    domain : tuple[float]
+        Domain over which samples were taken.
+    axis : int
+        Axis along which to transform.
+
+    Returns
+    -------
+    fq : jnp.ndarray
+        Real function value at query points.
+
+    """
+    # |a| << |xq|, so move a instead
+    a = (
+        jnp.moveaxis(a, axis, -1)
+        .at[..., 0]
+        .divide(2.0)
+        .at[..., -1]
+        .divide(1.0 + ((n % 2) == 0))
+    )
+    n = jnp.fft.rfftfreq(n, d=(domain[-1] - domain[0]) / (2 * jnp.pi * n))
+    basis = jnp.exp(1j * n * (xq - domain[0])[..., jnp.newaxis])
+    return 2.0 * (basis * a).real.sum(axis=-1)
+
+
+def interp_rfft2(
+    xq0, xq1, f, domain0=(0, 2 * jnp.pi), domain1=(0, 2 * jnp.pi), axes=(-2, -1)
+):
+    """Interpolate real-valued ``f`` to coordinates ``(xq0,xq1)`` with FFT.
+
+    Parameters
+    ----------
+    xq0 : jnp.ndarray
```

Comment:
not clear to me from the code/docstring, must xq0 and xq1 have the same length

Reply:
they just need to broadcast. works like any numpy function